app/scripts/modules/kubernetes/src/loadBalancer/details/details.controller.ts

```diff
@@ -29,25 +27,14 @@ class KubernetesLoadBalancerDetailsController implements IController {
     loadBalancer: ILoadBalancerFromStateParams,
     private app: Application,
   ) {
-    this.loadBalancerFromParams = loadBalancer;
-    this.app
-      .getDataSource('loadBalancers')
+    const dataSource = this.app.getDataSource('loadBalancers');
+    dataSource
       .ready()
       .then(() => {
-        this.extractLoadBalancer();
-        const unsubscribe = KubernetesManifestService.makeManifestRefresher(
-          this.app,
-          {
-            account: this.loadBalancerFromParams.accountId,
-            location: this.loadBalancerFromParams.region,
-            name: this.loadBalancerFromParams.name,
-          },
-          this,
-        );
-        this.$scope.$on('$destroy', () => {
-          unsubscribe();
-        });
-      });
+        this.extractLoadBalancer(loadBalancer);
+        dataSource.onRefresh(this.$scope, () => this.extractLoadBalancer(loadBalancer));
```

Comment:
This is mostly just out of curiosity, but is there no way to collapse these two calls into one?  It looks like everywhere we do X when the data source is ready, then also subscribe on refresh to do X again.  I guess the data source probably just doesn't expose a "do X when ready and every time it refreshes"...

Reply:
We have a [custom React hook](https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/presentation/hooks/useDataSource.hook.ts) for (I think) exactly that now! We can use them when we convert these details views to React, but would likely want to start by ensuring all leaf node components in each details template are React first. But yeah, great question, my understanding of one of the primary reasons for the introduction of React hooks is that the pattern of "fetch something once after first render and then set up a subscription or re-fetch in response to other things changing" in components written as classes had duplicated and convoluted logic in `componentDidMount` and `componentDidUpdate` methods, so they made a more first-class way with hooks. Better explanation [here](https://reactjs.org/docs/hooks-intro.html#complex-components-become-hard-to-understand) if you are interested.

Reply:
Thank you for the excellent explanation! 💯